### ISSUE TITLE
Promises with functions/filters?

### ISSUE THREAD

**Issue by ericclemmons** [twigjs/twig.js#61]
Perhaps this is the wrong place to put this, but I was curious about soliciting input on this request...

Suppose I have a route like `/:page` that will simply `res.render(page + '.twig')`.  These views utilized lots of API information from services like Github, Basecamp, etc.

Suppose I have in my templates:

``` twig
{{ github_api('total_issues') }}
```

Would it be possible, if `github_api` return s a Promise, for the rendering engine to continue compilation & processing once the callback is called?

When our designers & developers write templates with Twig (PHP), the synchronous nature allows lazy-loading of resources without the controllers having to expose dozens if not hundreds of data-points that are mostly not used.

With Twig.js, the asynchronous nature means that the immediate return value is only what's available, and these data points would have to be pre-fetched, which would mean hundreds of unnecessary API calls.

I'm willing to code this functionality in, but as evident in my last ticket & PR, I'd need someone to point me to the pertinent sections to adjust (otherwise I'll be floundering).

Thanks!

**Comment by Gertt** [comment]
Interested in this myself. We hacked together a super ugly workaround for an internal project at work, but it would be nice to have a way to let functions/filters preform async tasks.

I know it was mentioned in https://github.com/twigjs/twig.js/issues/357 that this is not available in PHP's Twig and therefor less important, but I'd also like to point out that PHP lets you do everything (web requests etcà synchronous while JS does not.

I might take a stab at providing a PR for this, but would like to know the maintainers' opinions on how best to implement this before starting out. Especially in regards to Promises vs callbacks etc. since this should also run in the browser and native promises are not supported yet everywhere. (I'd prefer Promise myself, personally though.)

**Comment by dave-irvine** [comment]
If it was going to be done with anything I would suggest Promises. As to how to achieve it I have absolutely no idea I'm afraid!

**Comment by Gertt** [comment]
After some looking around, I think it's doable. But if you want to maintain backwards compatibility, you'll probably end up with a complete mess in the code. I suppose the behaviour where `template.render` returns a synchronously rendered result is a must?

**Comment by dave-irvine** [comment]
I think we would need to support both sync and async rendering going forward.

Potentially we could split the twig.js package up so that the renderers are separate, and you would then

`npm install twig twig-async`

or

`npm install twig twig-sync`

Any thoughts?

**Comment by Gertt** [comment]
I am still looking into this and how exactly everything in the library fits together. The biggest problem I think will be existing custom tags.

If you would add async functionality to the library, it would mean that every part of the library needs to support it. For example, if I use the output of my custom async function as the input of another function or try to pass it through a filter, those things would also need to be able to handle the fact that my custom function returns its value in an async manner.

Now in most cases, I don't think this would be a problem. As far as I can tell, the only thing we would need to do is make sure that every part of the library that uses either `Twig.expression.parse` or `Twig.logic.parse` can handle that these could possible return a Promise. What's left then is -- depending on which render function was called -- returning a Promise that resolves to the rendered template or throw an error saying that the user is tryng to do async stuff with the sync render. 

However, when we start talking custom tags, the person who creates these tags is calling the above functions himself to evaluate the part of the template between the tags. Now this person becomes responsible for making sure that the tag is able to support async functionality. Once the documentation is updated, this shouldn't be a problem for future users. It's just when you start matching old custom tags that aren't async-ready with new async behaviour, you'll get in trouble.

Would like to hear your thoughts on this and possible mistakes in what I outlined above.

**Comment by Gertt** [comment]
Furthermore, like I said in a previous comment, supporting both at the same time will probably turn into a mess.

As an example, I have quickly whipped up an async implementation for just one expression: `Twig.expression.type.test`, which shows some of the implications of adding async support.

```
{
    type: Twig.expression.type.test,
    regex: /^is\s+(not)?\s*([a-zA-Z_][a-zA-Z0-9_]*(\s?as)?)/,
    next: Twig.expression.set.operations.concat([Twig.expression.type.parameter.start]),
    compile: function(token, stack, output) {
        token.filter   = token.match[2];
        token.modifier = token.match[1];
        delete token.match;
        delete token.value;
        output.push(token);
    },
    parse: function(token, stack, context) {
        var value = stack.pop(),
            params = token.params && Twig.expression.parse.apply(this, [token.params, context]),
            result = Twig.test(token.filter, value, params);

        if (token.modifier == 'not') {
            stack.push(!result);
        } else {
            stack.push(result);
        }
    },
    /**
     * Example of separate async function
     */
    parseAsync: function(token, stack, context) {
        var value = stack.pop(),
            params = token.params && Twig.expression.parse.apply(this, [token.params, context]);

        // both value or one of the params could've been async
        return Promise.all([ value, params ]).then(([ actualValue, actualParams])) => {
            var result = Twig.test(token.filter, actualValue, actualParams);

            if (token.modifier == 'not') {
                stack.push(!result);
            } else {
                stack.push(result);
            }
        });
    },
    /**
     * Example of combine sync / async function
     */
    parseAllInOne: function(token, stack, context) {
        var value = stack.pop(),
            params = token.params && Twig.expression.parse.apply(this, [token.params, context]),
            doTest = function(value, params) {
                var result = Twig.test(token.filter, value, params);

                if (token.modifier == 'not') {
                    stack.push(!result);
                } else {
                    stack.push(result);
                }
            };

        if (value.then || params.then) {
            return Promise.all([ value, params ]).then(([ actualValue, actualParams])) => {
                doTest(actualValue, actualParams);
            })
        } else {
            doTest(value, params);
        }
    },
},
```

As you can see in the `parseAsync` part, anything in the stack as well as anything returned by expression.parse could possibly be a Promise and this needs to be handled. `parseAllInOne` shows how you could combine everything into one function.

A solution like @dave-irvine suggested (creating a sync and async package) sounds like a good idea although we might end up with lots of duplicate code. Haven't really thought about the details for this.

**Comment by willrowe** [comment]
It is now possible to render both synchronously and asynchronously. If there is any specific functionality described in this issue that is still not possible, please open a new issue.
